src/main/java/blusunrize/immersiveengineering/common/items/BuzzsawItem.java

```diff
@@ -0,0 +1,775 @@
+	{
+		IItemHandler inv = stack.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null)
+				.orElseThrow(RuntimeException::new);
+		if(getUpgrades(stack).getBoolean("spareblades"))
+			return new Slot[]
+					{
+							new IESlot.WithPredicate(inv, 0, 88, 22, (itemStack) -> sawblades.contains(itemStack.getItem())),
+							new IESlot.Upgrades(container, inv, 1, 88, 52, "BUZZSAW", stack, true, getWorld),
+							new IESlot.Upgrades(container, inv, 2, 108, 52, "BUZZSAW", stack, true, getWorld),
+							new IESlot.WithPredicate(inv, 3, 108, 12, (itemStack) -> sawblades.contains(itemStack.getItem())),
+							new IESlot.WithPredicate(inv, 4, 108, 32, (itemStack) -> sawblades.contains(itemStack.getItem()))
+					};
+		else
+			return new Slot[]
+					{
+							new IESlot.WithPredicate(inv, 0, 98, 22, (itemStack) -> sawblades.contains(itemStack.getItem())),
+							new IESlot.Upgrades(container, inv, 1, 88, 52, "BUZZSAW", stack, true, getWorld),
+							new IESlot.Upgrades(container, inv, 2, 108, 52, "BUZZSAW", stack, true, getWorld)
+					};
+	}
+
+	@Override
+	public boolean canModify(ItemStack stack)
+	{
+		return true;
+	}
+
+	@Override
+	public void removeFromWorkbench(PlayerEntity player, ItemStack stack)
+	{
+		LazyOptional<IItemHandler> invCap = stack.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null);
+		invCap.ifPresent(inv -> {
+			if(!inv.getStackInSlot(0).isEmpty()&&!inv.getStackInSlot(1).isEmpty()&&!inv.getStackInSlot(2).isEmpty()&&!inv.getStackInSlot(3).isEmpty())
+				Utils.unlockIEAdvancement(player, "main/upgrade_buzzsaw");
+		});
+	}
+
+	@Override
+	public void recalculateUpgrades(ItemStack stack, World w)
+	{
+		super.recalculateUpgrades(stack, w);
+		FluidStack fs = getFluid(stack);
+		if(fs!=null&&fs.getAmount() > this.getCapacity(stack, 2000))
+		{
+			fs.setAmount(this.getCapacity(stack, 2000));
+			ItemNBTHelper.setFluidStack(stack, "Fluid", fs);
+		}
+	}
+
+	@Override
+	public void finishUpgradeRecalculation(ItemStack stack)
+	{
+		FluidStack fs = getFluid(stack);
+		if(fs!=null&&fs.getAmount() > getCapacity(stack, 2000))
+		{
+			fs.setAmount(getCapacity(stack, 2000));
+			ItemNBTHelper.setFluidStack(stack, "Fluid", fs);
+		}
+	}
+
+	public ItemStack getSawblade(ItemStack itemStack)
+	{
+		return getSawblade(itemStack, 0);
+	}
+
+	public ItemStack getSawblade(ItemStack itemStack, int spare)
+	{
+		if(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY==null)
+			return ItemStack.EMPTY;
+		ItemStack sawblade;
+		boolean remote = EffectiveSide.get()==LogicalSide.CLIENT;
+		LazyOptional<IItemHandler> cap = itemStack.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null);
+		// handle spares
+		int slot = spare==0?0: 2+spare;
+		String key = "sawblade"+(spare==0?"": ("_spare"+spare));
+		if(!remote&&cap.map(h -> h.getStackInSlot(slot).isEmpty()).orElse(false))
+			remote = true;
+		else if(remote&&!ItemNBTHelper.hasKey(itemStack, key))
```

Comment:
Doesn't this have the same issues with JEI as the revolver had recently?

Reply:
Hmm, I've not noticed the issue yet. I've spawned buzzsaws and drills from JEI and they worked fine